### ISSUE TITLE
Frontend preview for events not working

### ISSUE THREAD

**Issue by Defcon0** [contao/contao#2278]
**Affected version(s)**

tested in Contao 4.4

**Description**

To my knowledge it's possible to use the frontend preview to show events if their edit form is opened in contao.

I tried to click on frontend preview on a page like this:

```
https://www.acme.de/contao?do=calendar&table=tl_calendar_events&id=10414&act=edit&rt=bHsr8DhRP996ys_WX0p0b_vkHeNFxHoQalSAfcGW3c0&ref=SP02lwVd
```

If the domain I'm currently logged in in contao differs from the one specified in the calendar, the link of the frontend preview iframe is

```
https://www.acme.de/https://veranstaltung.corporate.acme.hhdev/de/anmeldung/some-alias
```

The domains explicitls in tl_page are set for both page roots.

Is this a bug?

Thanks in advance!

Bye Defcon0

**Comment by leofeyer** [comment]
This is because we are adding scheme and host to the URL even if `generateEventUrl()` already returned an absolute URL:

https://github.com/contao/contao/blob/e6f939884cfaecdcbab28c28d28b891751bc2821/calendar-bundle/src/EventListener/PreviewUrlConvertListener.php#L69

@contao/developers Would you consider it a BC break if we changed the `generateEventUrl()` method to always return an absolute URL? It already does if the current host does not match the target host.

**Comment by aschempp** [comment]
I think the listener should only add the scheme and host if the URL is not already absolute? Same as we do in `PageModel::generateAbsoluteUrl()`.

**Comment by leofeyer** [comment]
Unfortunately, the `$blnAbsoluteUrl` parameter for the `PageModel::generateAbsoluteUrl()` method is only available in Contao 4.9. Not sure if creating a version 4.4 only fix it is worth the effort, given that Contao 4.4 transitions into the last LTS phase by the end of this year.

**Comment by aschempp** [comment]
Thats not what I meant. `PageModel::getFrontendUrl` already generates absolute URLs if necessary, we just need to check for `http://` prefix before adding the base.

**Comment by leofeyer** [comment]
We have already fixed this exact issue in Contao 4.9 by adding the `$blnAbsoluteUrl` parameter, so whatever we decide here, it will be a Contao 4.4 only fix. 🤷🏻‍♂️

**Comment by Defcon0** [comment]
I understand but there are many existing contao 4.4 instances which are still and will be running for months/years :-(

**Comment by ausi** [comment]
I agree with @aschempp checking in https://github.com/contao/contao/blob/e6f939884cfaecdcbab28c28d28b891751bc2821/calendar-bundle/src/EventListener/PreviewUrlConvertListener.php#L69 for absolute URLs should be enough to fix this in 4.4 IMO:

```php
$url = $eventsAdapter->generateEventUrl($eventModel);
if (!preg_match('#^https?://#i') {
   $url = $request->getSchemeAndHttpHost().'/'.$url;
}
$event->setUrl($url);
```

**Comment by aschempp** [comment]
Don't forget about the network path in Contao 4.10. The check should be the same as in PageModel 😉

**Comment by leofeyer** [comment]
As I said, the issue has already been fixed in Contao 4.9 and there will never be a network path, so we do not have to account for that.

**Comment by leofeyer** [comment]
Since Contao 4.4 is now in the "security updates only" phase of the LTS cycle and the issue is already fixed in Contao 4.9, I am closing this ticket.
